subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/dependencysubstitution/DependencySubstitutionResolver.java

```diff
@@ -36,11 +37,13 @@ public DependencySubstitutionResolver(DependencyToComponentIdResolver resolver,
     public void resolve(DependencyMetadata dependency, BuildableComponentIdResolveResult result) {
         ComponentSelector selector = dependency.getSelector();
         DependencySubstitutionInternal details = new DefaultDependencySubstitution(selector, dependency.getRequested());
-        try {
-            rule.execute(details);
-        } catch (Throwable e) {
-            result.failed(new ModuleVersionResolveException(selector, e));
-            return;
+        synchronized (DependencySubstitutionResolver.class) {
```

Comment:
It would be even better if we used our project locking infrastructure here, so that substitution rules in different projects can run in parallel when `--parallel` is specified. The (not so well named) `WorkerLeaseService` has what we need for that.